### PR TITLE
refactor: move protobuf conversion logic to generated_types

### PR DESCRIPTION
**Pull request by tustvold**
Builds on top of #1428 and moves the protobuf conversions into the generated_types crate. This avoids crates depending on data_types from also depending on the entire protobuf + tonic + hyper + tower ecosystem.

It was largely a copy paste job but:

* The database_rules conversions are split into separate modules instead of a single large file in generated_types, I left the single large file in data_types alone
* I had to split one of the shard tests as it was testing both the sharder and the protobuf conversion

### CODE REVIEW

File: generated_types/src/google.rs [hunk 1]
```diff
@@ -300,3 +295,107 @@ impl From<QuotaFailure> for tonic::Status {
         )
     }
 }
+
```
Comment (tustvold) [inline generated_types/src/google.rs:298]: Moved from data_types::field_violation

File: data_types/src/database_rules.rs [hunk 2]
```diff
@@ -1206,514 +643,29 @@ mod tests {
...
     #[test]
+    #[allow(clippy::trivial_regex)]
     fn test_sharder() {
-        let protobuf = management::ShardConfig {
-            specific_targets: vec![management::MatcherToShard {
-                matcher: Some(management::Matcher {
-                    table_name_regex: "pu$".to_string(),
-                    ..Default::default()
-                }),
+        let shards: Vec<_> = (1000..1000000).collect();
```
Comment (tustvold) [inline data_types/src/database_rules.rs:649]: A test for the protobuf conversion alone is added to generated_types

File: generated_types/src/database_rules/shard.rs [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+
+        let shard_config: ShardConfig = protobuf.try_into().unwrap();
+
+        assert_eq!(shard_config.shards.len(), 2);
+        assert_eq!(shard_config.shards[&1].len(), 3);
+        assert_eq!(shard_config.shards[&2].len(), 1);
+    }
+
+    #[test]
+    fn test_sharder() {
```
Comment (tustvold) [inline generated_types/src/database_rules/shard.rs:343]: This was split out from the existing test in data_types

File: generated_types/src/database_rules.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    ValidationError(#[from] FieldViolation),
+}
+
+#[derive(Debug, Error)]
+pub enum EncodeError {
+    #[error("failed to encode protobuf: {0}")]
+    EncodeError(#[from] prost::EncodeError),
+}
+
+pub fn decode_database_rules(bytes: prost::bytes::Bytes) -> Result<DatabaseRules, DecodeError> {
```
Comment (tustvold) [inline generated_types/src/database_rules.rs:76]: These were previously member functions on DatabaseRules, this is no longer possible, but I think this is fine

File: generated_types/src/database_rules.rs [hunk 5]
```diff
@@ -0,0 +1,189 @@
+use std::convert::{TryFrom, TryInto};
+
+use thiserror::Error;
```
Comment (tustvold) [inline generated_types/src/database_rules.rs:3]: The clients use thiserror, not Snafu, so using it here for consistency - I also happen to prefer it so... If a problem can change

File: data_types/Cargo.toml [hunk 6]
```diff
@@ -8,15 +8,12 @@ readme = "README.md"
 
 [dependencies] # In alphabetical order
 chrono = { version = "0.4", features = ["serde"] }
-generated_types = { path = "../generated_types" }
```
Comment (alamb) [inline data_types/Cargo.toml:11]: 🎉

File: generated_types/src/lib.rs [hunk 7]
```diff
@@ -2,13 +2,6 @@
 // crates because of all the generated code it contains that we don't have much
 // control over.
 #![deny(broken_intra_doc_links)]
-#![allow(
```
Comment (alamb) [inline generated_types/src/lib.rs:5]: 👍  another win

**Review by alamb — APPROVED** [review]
Thank you for taking on this task -- I for one appreciate it greatly. 

I can't say I verified all lines for change during copy/paste but I did double check that all the [moved tests still still run (and pass)](https://app.circleci.com/pipelines/github/influxdata/influxdb_iox/5603/workflows/66c489db-b65e-4058-8ef5-97b852fef775/jobs/18363), and I have sufficient confidence in our coverage of end to end tests that I am pleased with this change.

🤞  here is hoping for better `rust_analyzer` support for generated types at some point. I find it hard to work in that crate as the analyzer has no idea what is going on.

However, this seems like a good step forward so I am 💯

**Pull request merged**